per-tenant-remote-sync: storage_sync metrics + upload retry unit tests

problame opened the pull request:
Curiously, the added unit test fails after attach with

```
              stdout: Starting existing postgres main_pg_node...
Safekeepers synced on 0/174DCF0
Extracting base backup to create postgres instance: path=/home/cs/src/neon-work-2/test_output/test_remote_storage_upload_queue_retries[local_fs]/repo/pgdatadirs/tenants/c5debc0d4332fb4c4068c8d7ed3ee6d7/main_pg_node port=15012

              stderr: command failed: extracting base backup failed

Caused by:
    0: failed to iterate over archive
    1: db error: ERROR: could not find data for key 010000000000000000000000000000000000 at LSN 0/174DCF1, for request at LSN 0/174DCF0
```

But it even fails if I disable the failpoints, so, it can't be related to the retry logic.

hlinnaka commented:
> Curiously, the added unit test fails after attach with ...

Do you mean `test_remote_storage_upload_queue_retries`? Weird, it passes on my laptop. `cargo test` also passes.

In the CI, it fails in a different way:
```
___ ERROR at teardown of test_remote_storage_upload_queue_retries[local_fs] ____
[gw0] linux -- Python 3.9.2 /github/home/.cache/pypoetry/virtualenvs/neon-_pxWMzVK-py3.9/bin/python
/github/home/.cache/pypoetry/virtualenvs/neon-_pxWMzVK-py3.9/lib/python3.9/site-packages/allure_commons/_allure.py:200: in __call__
    return self._fixture_function(*args, **kwargs)
test_runner/fixtures/neon_fixtures.py:1045: in neon_env_builder
    yield builder
test_runner/fixtures/neon_fixtures.py:787: in __exit__
    self.env.pageserver.assert_no_errors()
test_runner/fixtures/neon_fixtures.py:1836: in assert_no_errors
    assert not errors
E   assert not ['2022-11-22T11:18:54.125016Z ERROR compaction_loop{tenant_id=a4647354d59c4028241478a82f22ee32}:compact_timeline{timel...ut while waiting for WAL record at LSN 0/174D7B0 to arrive, last_record_lsn 0/16960E8 disk consistent LSN=0/16960E8\n"]
```

hlinnaka commented on pageserver/src/storage_sync.rs:
```diff
@@ -360,6 +364,19 @@ impl std::fmt::Display for UploadOp {
     }
 }
 
+macro_rules! upload_queue_items_metric {
+    ($remote_timeline_client:expr, $delta:expr) => {{
+        let remote_timeline_client = &$remote_timeline_client;
+        REMOTE_UPLOAD_QUEUE_UNFINISHED_TASKS
+                .get_metric_with_label_values(&[
+                    &remote_timeline_client.tenant_id.to_string(),
+                    &remote_timeline_client.timeline_id.to_string(),
+                ])
+                .unwrap()
+                .add($delta)
+    }}
+}
+
```
This could be a regular function, like this:

[upload_queue_items_metric.patch.txt](https://github.com/neondatabase/neon/files/10067399/upload_queue_items_metric.patch.txt)

hlinnaka replied:
I added a commit with this change to this PR

hlinnaka commented on test_runner/regress/test_remote_storage.py:
```diff
@@ -97,6 +98,15 @@ def test_remote_storage_backup_and_restore(
         log.info(f"waiting for checkpoint {checkpoint_number} upload")
+
+        # insert upload failpoints to exercise retry code path
+        action = "3*return->off" if checkpoint_number == 0 else "off"
```
The checkpoints are numbered 1 and 2, see line 82 above:
```
checkpoint_numbers = range(1, 3)
```
Therefore, I believe this `if checkpoint_number == 0` condition is never true.

Also, isn't it too late to set the failpoint, after the checkpoint has already been started?

I'd suggest moving this configure_failpoints call to before the loop

hlinnaka replied:
Added a commit with that change to this PR

hlinnaka commented:
> > Curiously, the added unit test fails after attach with ...
> 
> Do you mean `test_remote_storage_upload_queue_retries`? 

Ok, after fixing some other issues, I'm now seeing the error that you reported, and the CI is now failing on that too.

hlinnaka commented:
This snippet in the log seems suspicious:

```
2022-11-22T14:17:14.596817Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}: starting on 1 timelines
2022-11-22T14:17:14.624087Z  INFO layer flush task{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:flush_frozen_layer{tenant_id=ac6e5888034d35533aa5153270665238 timeline_id=ec5a4ec5775248bc744f07f52a7695a4 layer=inmem-0000000001708EB9-00000000017503A1}: scheduled layer file upload /home/heikki/git-sandbox/neon/test_output/test_remote_storage_upload_queue_retries[local_fs]/repo/tenants/ac6e5888034d35533aa5153270665238/timelines/ec5a4ec5775248bc744f07f52a7695a4/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001708EB9-00000000017503A1
2022-11-22T14:17:14.624149Z  INFO layer flush task{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:flush_frozen_layer{tenant_id=ac6e5888034d35533aa5153270665238 timeline_id=ec5a4ec5775248bc744f07f52a7695a4 layer=inmem-0000000001708EB9-00000000017503A1}: scheduled metadata upload with 1 files
2022-11-22T14:17:14.635137Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:checkpoint{cconf=Forced tenant_id=ac6e5888034d35533aa5153270665238 timeline_id=ec5a4ec5775248bc744f07f52a7695a4}: Starting Level0 compaction in LSN range 0/1708EB9-0/17503A1 for 1 layers (1 deltas in total)
2022-11-22T14:17:14.635202Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:checkpoint{cconf=Forced tenant_id=ac6e5888034d35533aa5153270665238 timeline_id=ec5a4ec5775248bc744f07f52a7695a4}: compact includes 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001708EB9-00000000017503A1
2022-11-22T14:17:14.661287Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:checkpoint{cconf=Forced tenant_id=ac6e5888034d35533aa5153270665238 timeline_id=ec5a4ec5775248bc744f07f52a7695a4}: scheduled layer file deletion /home/heikki/git-sandbox/neon/test_output/test_remote_storage_upload_queue_retries[local_fs]/repo/tenants/ac6e5888034d35533aa5153270665238/timelines/ec5a4ec5775248bc744f07f52a7695a4/000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001708EB9-00000000017503A1
2022-11-22T14:17:14.661380Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}: timeline ec5a4ec5775248bc744f07f52a7695a4 checkpoint_before_gc done
2022-11-22T14:17:14.668886Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:gc_timeline{timeline=ec5a4ec5775248bc744f07f52a7695a4 cutoff=0/17503A0}: scheduled metadata upload with 0 files
2022-11-22T14:17:14.668936Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:gc_timeline{timeline=ec5a4ec5775248bc744f07f52a7695a4 cutoff=0/17503A0}: GC starting
2022-11-22T14:17:14.668996Z  INFO manual_gc{tenant=ac6e5888034d35533aa5153270665238 timeline=ec5a4ec5775248bc744f07f52a7695a4}:gc_timeline{timeline=ec5a4ec5775248bc744f07f52a7695a4 cutoff=0/17503A0}: GC completed removing 0 layers, cutoff 0/17503A0
2022-11-22T14:17:14.669109Z  INFO PUT /v1/tenant/ac6e5888034d35533aa5153270665238/timeline/ec5a4ec5775248bc744f07f52a7695a4/do_gc 200 OK
```

So a new delta layer `000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001708EB9-00000000017503A1` is created and scheduled for upload. Right after that, compaction chooses that file for compaction and schedules its deletion.

The compaction should have created some L1 files to replace `000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000001708EB9-00000000017503A1`, and scheduled their upload before scheduling the deletion. Why did that not happen?

problame commented:
I added file_kind and op kind labels to the number-of-unfinished tasks metric, like we have for the duration metric. Also reworked the metric-update-function to take a `&Op` and infer the label values in a `match`.

The test now uses the metric to ensure that we're queuing & retrying all three kinds of upload tasks.

@hlinnaka @LizardWizzard this good enough for merge?

hlinnaka commented:
> @hlinnaka @LizardWizzard this good enough for merge?

Yep, ship it!
